### PR TITLE
expression: wrong result of timestampadd(month,1,date '2024-01-31')

### LINKED ISSUE #41052: wrong result of timestampadd(month,1,date '2024-01-31')
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

The issue if from https://ask.pingcap.com/t/timestampadd-function/419.

```sql
select timestampadd(month,1,date '2024-01-31');
```

### 2. What did you expect to see? (Required)

MySQL result is expected, as following:
```sql
> select timestampadd(month,1,date '2024-01-31');
+-----------------------------------------+
| timestampadd(month,1,date '2024-01-31') |
+-----------------------------------------+
| 2024-02-29                              |
+-----------------------------------------+
```

### 3. What did you see instead (Required)

```sql
> select timestampadd(month,1,date '2024-01-31');
+-----------------------------------------+
| timestampadd(month,1,date '2024-01-31') |
+-----------------------------------------+
| 2024-03-02                              |
+-----------------------------------------+
```

### 4. What is your TiDB version? (Required)

```sql
***************************[ 1. row ]***************************
tidb_version() | Release Version: v6.6.0-alpha-346-gd0d321f440
Edition: Community
Git Commit Hash: d0d321f440e265e6eb87731ba9fbf3611bf0dca2
Git Branch: master
UTC Build Time: 2023-02-04 00:17:06
GoVersion: go1.19.3
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: unistore
```

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #41052

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/expression/builtin_time.go [hunk 1]
```diff
@@ -6232,6 +6232,11 @@ func addUnitToTime(unit string, t time.Time, v float64) (time.Time, bool, error)
 			return tb, true, nil
 		}
 		tb = t.AddDate(0, int(v), 0)
+
+		// For corner case: timestampadd(month,1,date '2024-01-31') = "2024-02-29", timestampadd(month,1,date '2024-01-30') = "2024-02-29"
+		for tb.Month() != t.Month()+1 {
```
Comment (yibin87) [inline pkg/expression/builtin_time.go:6237]: What if add more than 1 month?

Comment (xzhangxian1008) [reply pkg/expression/builtin_time.go:6237]: > What happen if adding more than 1 month?

This is mysql: timestampadd(month,1,date '2024-01-31') = "2024-02-29"

This is tidb: timestampadd(month,1,date '2024-01-31') = "2024-03-02"

The month field of the result of `timestampadd(month,1,date '2024-01-31')` is always 1 more than the month field of input. Adding more than 1 month is unexpected.

File: pkg/expression/builtin_time.go [hunk 2]
```diff
@@ -6232,6 +6232,12 @@ func addUnitToTime(unit string, t time.Time, v float64) (time.Time, bool, error)
 			return tb, true, nil
 		}
 		tb = t.AddDate(0, int(v), 0)
+
+		// For corner case: timestampadd(month,1,date '2024-01-31') = "2024-02-29", timestampadd(month,1,date '2024-01-30') = "2024-02-29"
+		for int(tb.Month())%12 != (int(t.Month())+int(v))%12 {
+			time.Sleep(10 * time.Millisecond)
```
Comment (yibin87) [inline pkg/expression/builtin_time.go:6238]: why need to sleep here?

**Review by yibin87 — APPROVED** [review]
Others LGTM

**Comment by SeaRise** [comment]
/cc @XuHuaiyu @windtalker 
Need a LGTM from the expression owners.

File: pkg/expression/builtin_time.go (hunk 2 above)
Comment (xzhangxian1008) [reply pkg/expression/builtin_time.go:6238]: > why need to sleep here?

mistake, I will delete it.

**Review by windtalker — APPROVED** [review]
LGTM

**Review by XuHuaiyu — APPROVED** [review]

File: pkg/expression/builtin_time_test.go [hunk 3]
```diff
@@ -2580,6 +2580,18 @@ func TestTimestampAdd(t *testing.T) {
 		{"MINUTE", 1.5, "1995-05-01 00:00:00", "1995-05-01 00:02:00"},
 		{"MINUTE", 1.5, "1995-05-01 00:00:00.000000", "1995-05-01 00:02:00"},
 		{"MICROSECOND", -100, "1995-05-01 00:00:00.0001", "1995-05-01 00:00:00"},
+
+		// issue41052
+		{"MONTH", 1, "2024-01-31", "2024-02-29 00:00:00"},
```
Comment (zanmato1984) [inline pkg/expression/builtin_time_test.go:2585]: We should've at least had one case for negative month.

Comment (xzhangxian1008) [reply pkg/expression/builtin_time_test.go:2585]: > We should've at least had one case for negative month.

fixed in https://github.com/pingcap/tidb/pull/54916

**Pull request merged**